open api codegen produces faulty constructor

### Expected Behavior

No compilation problems when generating micronaut clients and pojo's produced when referencing a valid open api file

### Actual Behaviour

an invalid constructor is created (`public CategoryObject(, String locale, String name)`)

gradle output: 
```sh
> Task :compileJava FAILED
C:\Users\jonat\src\git.pbu.lol\zendesk4j\build\generated\openapi\generateClientOpenApiModels\src\main\java\lol\pbu\zendesk\model\CategoryObject.java:145: error: illegal start of type
    public CategoryObject(, String locale, String name) {
                          ^
1 error

[Incubating] Problems report is available at: file:///C:/Users/jonat/src/git.pbu.lol/zendesk4j/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler output below.
  C:\Users\jonat\src\git.pbu.lol\zendesk4j\build\generated\openapi\generateClientOpenApiModels\src\main\java\lol\pbu\zendesk\model\CategoryObject.java:145: error: illegal start of type
      public CategoryObject(, String locale, String name) {
                            ^
  1 error

* Try:
> Check your code and dependencies to fix the compilation error(s)
```

### Steps To Reproduce

1. download [open api file](https://developer.zendesk.com/help_center/oas.yaml) from [zendesk docs page](https://developer.zendesk.com/api-reference/help_center/help-center-api/introduction/)
2. download a kotlin starter from micronaut launch whose `micronaut-cli.yml` contents provide this: (this is the state of the repo at [this commit](https://github.com/PeanutButter-Unicorn/zendesk4j/commit/807b50e052d2b79c6f4ae812b8094c54e5ab7f80))
```sh
applicationType: default
defaultPackage: lol.pbu
testFramework: spock
sourceLanguage: kotlin
buildTool: gradle_kotlin
features: [app-name, asciidoctor, data, data-jdbc, gradle, h2, http-client, jdbc-hikari, jul-to-slf4j, kotlin, kotlin-application, ksp, logback, micronaut-aot, micronaut-build, micronaut-http-validation, netty-server, readme, serialization-jackson, shade, spock, static-resources, testcontainers, yaml, yaml-build]
```
3. add the micronaut api gradle plugin & point the openapi configuration to the downloaded file from step 1, then assign the api and model package names
```groovy
plugins {
. . .
    id("io.micronaut.openapi") version "4.5.3"
. . .
}
. . .
micronaut {
  . . .
    openapi {
        client(file("src/docs/asciidoc/oas.yaml")) {
            apiPackageName = "lol.pbu.zendesk.api"
            modelPackageName = "lol.pbu.zendesk.model"
        }
    }
}
```
4. attempt to generate the client and models with `./gradlew compileJava --console=verbose`

### Environment Information

- Operating System: Windows 11
- JDK: 
<ol>

```sh
openjdk 21.0.7 2025-04-15 LTS
OpenJDK Runtime Environment Microsoft-11369940 (build 21.0.7+6-LTS)
OpenJDK 64-Bit Server VM Microsoft-11369940 (build 21.0.7+6-LTS, mixed mode, sharing)
```
</ol>

- gradle wrapper info
<ol>

```
------------------------------------------------------------
Gradle 8.13
------------------------------------------------------------

Build time:    2025-02-25 09:22:14 UTC
Revision:      073314332697ba45c16c0a0ce1891fa6794179ff

Kotlin:        2.0.21
Groovy:        3.0.22
Ant:           Apache Ant(TM) version 1.10.15 compiled on August 25 2024
Launcher JVM:  21.0.7 (Microsoft 21.0.7+6-LTS)
Daemon JVM:    C:\Program Files\Microsoft\jdk-21.0.7.6-hotspot (no JDK specified, using current Java home)
OS:            Windows 11 10.0 amd64
```
</ol>

### Example Application

https://github.com/PeanutButter-Unicorn/zendesk4j/commit/5d1aac3fb80ba1039460df4e5368e397e8093d60

### Version

4.8.2

--- a/openapi-generator/src/main/java/io/micronaut/openapi/generator/AbstractMicronautJavaCodegen.java
+++ b/openapi-generator/src/main/java/io/micronaut/openapi/generator/AbstractMicronautJavaCodegen.java
@@ -2032,6 +2032,16 @@ public abstract class AbstractMicronautJavaCodegen<T extends GeneratorOptionsBui
             }
             if (!requiredVarsWithoutDiscriminator.isEmpty()) {
                 model.vendorExtensions.put("requiredVarsWithoutDiscriminator", requiredVarsWithoutDiscriminator);
+                var requiredVarsWithoutDiscriminatorAndReadOnly = new ArrayList<CodegenProperty>();
+                for (var prop : requiredVarsWithoutDiscriminator) {
+                    if (!isServer && prop.isReadOnly) {
+                        continue;
+                    }
+                    requiredVarsWithoutDiscriminatorAndReadOnly.add(prop);
+                }
+                if (!requiredVarsWithoutDiscriminatorAndReadOnly.isEmpty()) {
+                    model.vendorExtensions.put("requiredVarsWithoutDiscriminatorAndReadOnly", requiredVarsWithoutDiscriminatorAndReadOnly);
+                }
             }
             model.allVars = allVars;
             model.vendorExtensions.put("requiredVars", requiredVars);

--- a/openapi-generator/src/main/resources/templates/java-micronaut/common/model/pojo.mustache
+++ b/openapi-generator/src/main/resources/templates/java-micronaut/common/model/pojo.mustache
@@ -192,7 +192,7 @@ public {{>common/model/sealed}}class {{classname}}{{#parent}} extends {{{parent}
         {{/micronaut_serde_jackson}}
         {{#formatNoEmptyLines}}
         {{#vendorExtensions.withRequiredVars}}
-    public {{classname}}({{#vendorExtensions.requiredVarsWithoutDiscriminator}}{{#isReadOnly}}{{#vendorExtensions.isServer}}{{^-first}}, {{/-first}}{{{vendorExtensions.typeWithEnumWithGenericAnnotations}}} {{name}}{{/vendorExtensions.isServer}}{{/isReadOnly}}{{^isReadOnly}}{{^-first}}, {{/-first}}{{{vendorExtensions.typeWithEnumWithGenericAnnotations}}} {{name}}{{/isReadOnly}}{{/vendorExtensions.requiredVarsWithoutDiscriminator}}) {
+    public {{classname}}({{#vendorExtensions.requiredVarsWithoutDiscriminatorAndReadOnly}}{{^-first}}, {{/-first}}{{{vendorExtensions.typeWithEnumWithGenericAnnotations}}} {{name}}{{/vendorExtensions.requiredVarsWithoutDiscriminatorAndReadOnly}}) {
         {{#parent}}
         {{#vendorExtensions.requiredParentVarsWithoutDiscriminator.0}}super({{/vendorExtensions.requiredParentVarsWithoutDiscriminator.0}}{{#vendorExtensions.requiredParentVarsWithoutDiscriminator}}{{#vendorExtensions.isServerOrNotReadOnly}}{{^-first}}, {{/-first}}{{name}}{{/vendorExtensions.isServerOrNotReadOnly}}{{/vendorExtensions.requiredParentVarsWithoutDiscriminator}}{{#vendorExtensions.requiredParentVarsWithoutDiscriminator.0}});{{/vendorExtensions.requiredParentVarsWithoutDiscriminator.0}}
         {{/parent}}

--- a/openapi-generator/src/test/java/io/micronaut/openapi/generator/JavaMicronautClientCodegenTest.java
+++ b/openapi-generator/src/test/java/io/micronaut/openapi/generator/JavaMicronautClientCodegenTest.java
@@ -1681,4 +1681,18 @@ class JavaMicronautClientCodegenTest extends AbstractMicronautCodegenTest {
                 }
             """);
     }
+
+    @Test
+    void testReadOnlyRequiredPropertyInConstructor() {
+        var codegen = new JavaMicronautClientCodegen();
+        String outputPath = generateFiles(codegen, "src/test/resources/3_0/oas.yml", CodegenConstants.MODELS, CodegenConstants.APIS);
+        String path = outputPath + "src/main/java/org/openapitools/";
+
+        assertFileContains(path + "model/CategoryObject.java", """
+                public CategoryObject(String locale, String name) {
+                    this.locale = locale;
+                    this.name = name;
+                }
+            """);
+    }
 }

--- a/openapi-generator/src/test/java/io/micronaut/openapi/generator/JavaMicronautServerCodegenTest.java
+++ b/openapi-generator/src/test/java/io/micronaut/openapi/generator/JavaMicronautServerCodegenTest.java
@@ -760,7 +760,7 @@ class JavaMicronautServerCodegenTest extends AbstractMicronautCodegenTest {
             """);
         assertFileContains(path + "model/Result.java", """
                 private String id;
-
+            
                 @Nullable(inherited = true)
                 @Schema(name = "date", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
                 @JsonProperty(JSON_PROPERTY_DATE)
@@ -779,5 +779,20 @@ class JavaMicronautServerCodegenTest extends AbstractMicronautCodegenTest {
         assertFileContains(path + "model/BookInfo.java",
             "import java.time.ZonedDateTime;",
             "private ZonedDateTime createdAt;");
+    }
+
+    @Test
+    void testReadOnlyRequiredPropertyInConstructor() {
+        var codegen = new JavaMicronautServerCodegen();
+        String outputPath = generateFiles(codegen, "src/test/resources/3_0/oas.yml", CodegenConstants.MODELS, CodegenConstants.APIS);
+        String path = outputPath + "src/main/java/org/openapitools/";
+
+        assertFileContains(path + "model/CategoryObject.java", """
+                public CategoryObject(Integer id, String locale, String name) {
+                    this.id = id;
+                    this.locale = locale;
+                    this.name = name;
+                }
+            """);
     }
 }

--- a/openapi-generator/src/test/resources/3_0/oas.yml
+++ b/openapi-generator/src/test/resources/3_0/oas.yml
@@ -1,0 +1,182 @@
+openapi: 3.0.2
+info:
+  title: Help Center API
+  description: Help Center v2 REST API
+  version: 2.0.0
+tags:
+  - name: Articles
+  - name: Article Labels
+  - name: Article Attachments
+  - name: Translations
+  - name: Categories
+  - name: Sections
+  - name: Sessions
+  - name: Votes
+  - name: User Segments
+  - name: Article Comments
+  - name: Search
+  - name: Topics
+  - name: Posts
+  - name: Post Comments
+  - name: Content Subscriptions
+  - name: User Subscriptions
+  - name: User Images
+paths:
+  /api/v2/help_center/{locale}/categories:
+    get:
+      operationId: ListCategories
+      tags:
+        - Categories
+      summary: List Categories
+      description: |
+        **Note**: `{/locale}` is an optional parameter for admins and agents. End users and anonymous users must provide the parameter.
+
+        #### Allowed for
+
+        * Anonymous users
+
+        The response will list only the categories that the agent, end user, or
+        anonymous user can view in the help center.
+
+        #### Pagination
+
+        * Cursor pagination (recommended)
+        * Offset pagination
+
+        See [Pagination](/api-reference/introduction/pagination/).
+
+        #### Sorting
+
+        You can sort the results with the `sort_by` and `sort_order` query string parameters.
+
+        ```
+        GET /api/v2/help_center/en-us/categories.json?sort_by=updated_at&sort_order=asc
+        ```
+
+        The `sort_by` parameter can have one of the following values:
+
+        | value         | description
+        | ------------- | -----------
+        | `position`    | order set manually using the Arrange Content page. Default order
+        | `created_at`  | order by creation time
+        | `updated_at`  | order by update time
+
+        The `sort_order` parameter can have one of the following values:
+
+        | value   | description
+        | ------- | -----------
+        | `asc`   | ascending order
+        | `desc`  | descending order
+
+        #### Sideloads
+
+        The following sideloads are supported:
+
+        | Name          | Will sideload
+        |---------------|--------------
+        | translations  | the category translations, if any
+
+        Translations are embedded within the category because they're
+        not shared between resources.
+      parameters:
+        - name: sort_by
+          in: query
+          description: Sorts the results by one of the accepted values
+          schema:
+            type: string
+            enum:
+              - position
+              - created_at
+              - updated_at
+        - name: sort_order
+          in: query
+          description: Selects the order of the results.
+          schema:
+            type: string
+            enum:
+              - asc
+              - desc
+      responses:
+        "200":
+          description: description
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/CategoriesResponse'
+components:
+  schemas:
+    CategoriesResponse:
+      type: object
+      properties:
+        categories:
+          type: array
+          items:
+            $ref: '#/components/schemas/CategoryObject'
+    CategoryObject:
+      type: object
+      properties:
+        created_at:
+          type: string
+          format: date-time
+          description: The time at which the category was created
+          readOnly: true
+        description:
+          type: string
+          description: The description of the category
+        html_url:
+          type: string
+          description: The url of this category in Help Center
+          readOnly: true
+        id:
+          type: integer
+          description: Automatically assigned when creating categories
+          readOnly: true
+        locale:
+          type: string
+          description: The locale where the category is displayed
+        name:
+          type: string
+          description: The name of the category
+        outdated:
+          type: boolean
+          description: Whether the category is out of date
+          readOnly: true
+        position:
+          type: integer
+          description: The position of this category relative to other categories
+        source_locale:
+          type: string
+          description: The source (default) locale of the category
+          readOnly: true
+        updated_at:
+          type: string
+          format: date-time
+          description: The time at which the category was last updated
+          readOnly: true
+        url:
+          type: string
+          description: The API url of this category
+          readOnly: true
+      example:
+        description: ""
+        html_url: https://company.zendesk.com/hc/en-us/categories/354362577
+        id: 1635
+        locale: en-us
+        name: Self Help Articles
+        source_locale: en-us
+        url: https://company.zendesk.com/api/v2/help_center/categories/354362577
+      required:
+        - id
+        - name
+        - locale
+    CategoryResponse:
+      type: object
+      properties:
+        category:
+          $ref: '#/components/schemas/CategoryObject'
+  securitySchemes:
+    basicAuth:
+      type: http
+      scheme: basic
+security:
+  - basicAuth: []

